materialize-starburst/starburst.go

```diff
@@ -0,0 +1,430 @@
+	if err != nil {
+		return nil, fmt.Errorf("initialization of file processor failed: %w", err)
+	}
+	defer storeFileProcessor.Destroy()
+
+	for it.Next() {
+		var b = t.bindings[it.Binding]
+		doc, err := b.target.Document.MappedType.Converter(it.RawJSON)
+		if err != nil {
+			return nil, fmt.Errorf("converting document %s: %w", b.target.Document.Field, err)
+		}
+		row := append(append(it.Key, it.Values...), doc)
+		//Store data to local file
+		if err := storeFileProcessor.Store(it.Binding, row); err != nil {
+			return nil, fmt.Errorf("storing file locally failed %w", err)
+		}
+	}
+
+	// Upload local files to cloud
+	_, err = storeFileProcessor.Commit()
+	if err != nil {
+		return nil, fmt.Errorf("storing file on cloud failed %w", err)
+	}
+
+	return func(ctx context.Context, runtimeCheckpoint *protocol.Checkpoint) (*pf.ConnectorState, m.OpFuture) {
+		var cp = checkpoint{}
+		for _, b := range t.bindings {
+			cpBinding := checkpointBinding{DataDirPrefix: storeFileProcessor.GetCloudPrefix(b.target.Binding),
+				CreateSql: b.store.createTempTable,
+				MergeSql:  b.store.mergeIntoTarget,
+				DropSql:   b.store.dropTempTable}
+			cp.QueriesSpec = append(cp.QueriesSpec, cpBinding)
+		}
+		t.cp = cp
+		var checkpointJSON, err = json.Marshal(cp)
+		if err != nil {
+			return nil, pf.FinishedOperation(fmt.Errorf("creating checkpoint json: %w", err))
+		}
+
+		t.store.stored = it.Total
+		return &pf.ConnectorState{UpdatedJson: checkpointJSON}, nil
+	}, nil
+}
+
+func (t *transactor) Acknowledge(ctx context.Context) (*pf.ConnectorState, error) {
+	return sql.AcknowledgeWithDelay(ctx, t.store.round, t.updateDelay, t.store.stored, func(ctx context.Context) (*pf.ConnectorState, error) {
+		log.Info("store: starting committing changes")
+		fileProcessor, err := NewParquetFileProcessor(ctx, t.s3Operator, nil, t.bindings, t.cfg.DataStoreDir)
+		if err != nil {
+			return nil, fmt.Errorf("initialization of file processor failed: %w", err)
+		}
+		defer fileProcessor.Destroy()
+		for _, qs := range t.cp.QueriesSpec {
+			dataDir := fmt.Sprintf("s3://%s/%s", t.cfg.Bucket, qs.DataDirPrefix)
+			// Create a binding-scoped temporary table for store documents to be merged into target table
+			if _, err := t.store.conn.ExecContext(ctx, qs.CreateSql, dataDir); err != nil {
+				return nil, fmt.Errorf("creating temp table fialed: %w", err)
+			}
+			// Merging temp table with target
+			if _, err := t.store.conn.ExecContext(ctx, qs.MergeSql); err != nil {
+				return nil, fmt.Errorf("merging fialed: %w", err)
+			}
+			// Drop temp table
+			if _, err := t.store.conn.ExecContext(ctx, qs.DropSql); err != nil {
+				return nil, fmt.Errorf("droping temp table fialed: %w", err)
+			}
+			//Clean cloud files
+			if err := fileProcessor.Delete(qs.DataDirPrefix); err != nil {
+				return nil, fmt.Errorf("deleting files on cloud failed: %w", err)
+			}
+		}
+		log.Info("store: finished committing changes")
+		return nil, nil
```

Comment:
nit: this isn't strictly required for correctness, but it's good practice to emit a ConnectorState update that "cleans up" the state checkpoint to reflect that these files have been applied.

That will prevent a re-attempted application done by a future connector invocation, which could happen in the relatively distant future (perhaps after cloud files have been removed by a bucket policy, for example).

Reply:
Here's an example of doing this [for Snowflake](https://github.com/estuary/connectors/blob/main/materialize-snowflake/snowflake.go#L557-L573). materialize-databricks also does this.